### ISSUE TITLE
How can we implement spectral normalization for GANs in Keras?

### ISSUE THREAD

**Issue by Hhhhhhhhhhao** [keras-team/keras#10405]
Hi, I'm currently thinking about using keras to implement the spectral normalization for GANs. I'm thinking to use the kernel_constraint, but it seems like only several available built-in functions can be used as kernel_constraint. Can I use my own functions as constraints?
Can someone give some ideas or instructions about this?

**Comment by Hhhhhhhhhhao** [comment]
Here is a implementation of mine, I didn't test it with my model.
`from keras.constraints import Constraint
from keras import backend as K

class spectral_norm(Constraint):
    def __init__(self, iteration=1):
        self.iteration = iteration

    def __call__(self, w):
        w_shape = w.shape.as_list()
        w = K.reshape(w, [-1, w_shape[-1]])

        u = K.get_variable("u", [1, w_shape[-1]], initializer=K.truncated_normal_initializer(), trainable=False)

        u_hat = u
        v_hat = None
        for i in range(self.iteration):
            # power iteration
            v_ = K.matmul(u_hat, K.transpose(w))
            v_hat = self.l2_norm(v_)

            u_ = K.matmul(v_hat, w)
            u_hat = self.l2_norm(u_)

        sigma = K.matmul(K.matmul(v_hat, w), K.transpose(u_hat))
        w_norm = w / sigma

        with K.control_dependencies([u.assign(u_hat)]):
            w_norm = K.reshape(w_norm, w_shape)

        return w_norm

    @staticmethod
    def l2_norm(v, eps=1e-12):
        return v / (K.reduce_sum(v ** 2) ** 0.5 + eps)`

most of the code are from here: https://github.com/taki0112/Self-Attention-GAN-Tensorflow/blob/master/ops.py

Hope someone can give me some advice on this.
